src/Autofac/RegistrationExtensions.Delegates.cs

```diff
@@ -0,0 +1,817 @@
+﻿// Copyright (c) Autofac Project. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Globalization;
+using System.Linq.Expressions;
+using System.Reflection;
+using Autofac.Builder;
+using Autofac.Core;
+using Autofac.Core.Activators.Delegate;
+using Autofac.Core.Activators.ProvidedInstance;
+using Autofac.Core.Activators.Reflection;
+using Autofac.Core.Lifetime;
+using Autofac.Core.Resolving.Pipeline;
+using Autofac.Features.Scanning;
+using Autofac.Util;
+
+namespace Autofac;
+
+/// <summary>
+/// Adds registration syntax to the <see cref="ContainerBuilder"/> type.
+/// </summary>
+[SuppressMessage("Microsoft.Maintainability", "CA1506:AvoidExcessiveClassCoupling")]
+public static partial class RegistrationExtensions
+{
+    /// <summary>
+    /// Register a delegate as a component.
+    /// </summary>
+    /// <typeparam name="TDependency1">The type of a dependency to inject into the delegate.</typeparam>
+    /// <typeparam name="TComponent">The type of the instance.</typeparam>
+    /// <param name="builder">Container builder.</param>
+    /// <param name="delegate">The delegate to register.</param>
+    /// <returns>Registration builder allowing the registration to be configured.</returns>
+    public static IRegistrationBuilder<TComponent, SimpleActivatorData, SingleRegistrationStyle>
+        Register<TDependency1, TComponent>(
+            this ContainerBuilder builder,
+            Func<IComponentContext, TDependency1, TComponent> @delegate)
+        where TDependency1 : notnull
+        where TComponent : notnull
+    {
+        if (@delegate == null)
+        {
+            throw new ArgumentNullException(nameof(@delegate));
+        }
+
+        return builder.Register((c, p) => @delegate(
```

Comment:
As I read this I wonder if it's worth doing anything with the parameters in this new way. Working with parameters is sort of a pain, too. Like, it'd be cool to do this:

```c#
builder.Register((IDependency1 dep) => new Thing(dep));
```

And also have it work with this:

```c#
container.Resolve<Thing>(TypedParameter.From(myDep));
```

That is, right now the dependencies are _only_ resolved, where a full resolution operation also generally takes parameters into account.

That would obviously impact perf in a way such that the majority case (no parameters) would always be checking for something that isn't there and then falling back to do the right thing here. Probably something "opt-in" would be best if that feature were to be added.

```c#
builder.Register(
  (IDependency1 dep1) => new Thing(dep1),
  includeParameters: true);
```

Something like that maybe.

Reply:
Oof; yeah, I see what you mean...and it would make it more consistent. Not super-straightforward though, but I'll take a look. There's already another overload that takes `IEnumerable<Parameter>()`.  

What I sort of want to happen (in a general sense) is that if I call `Resolve<TSomething>()` on a component context with parameters in it, that it looks at those parameters for a matching typed parameter first...but that would clash with the reflection activator, and wouldn't handle named delegate parameters. Hmm.

Reply:
Well, the parameters only go to the thing being activated, so in this case it'd be the `Thing`. I imagine it'd be something like:

```c#
public static IRegistrationBuilder<TComponent, SimpleActivatorData, SingleRegistrationStyle>
  Register<TDependency1, TComponent>(
    this ContainerBuilder builder,
    Func<TDependency1, TComponent> @delegate,
    bool includeParameters = false)
  where TDependency1 : notnull
  where TComponent : notnull
{
  if (@delegate == null)
  {
      throw new ArgumentNullException(nameof(@delegate));
  }

  return builder.Register((c, p) => 
  {
    TDependency1 dep1;
    if(includeParameters && p.FirstOrDefault(param => param.CanSupplyValue(/*....*/, out valueProvider)) is Parameter val)
    {
      dep1 = valueProvider();
    }
    else
    {
      dep1 = c.Resolve<TDependency1>();
    }

    @delegate(dep1);
}
```

Doing that from memory, not compiler. But that's the basic concept; Ideally the bool check wouldn't cost much in the majority case but folks could opt in to supporting parameters.

I suppose it could default to always check the list of parameters, but that wouldn't be as fast if you know you're never calling `Resolve<T>(params)`.

Reply:
OK; I've been having a bit of an experiment this morning, and I think I've got a good solution for handing parameters that doesn't require an additional flag, and doesn't detract from the performance.

Here's the benchmark code:

```csharp
[GlobalSetup]
public void Setup()
{
    var builder = new ContainerBuilder();

    builder.RegisterType<MyDependency1>();

    // Existing method, capture the argument directly.
    builder.Register((MyDependency1 dep1) => new MyComponent(dep1, "arg"))
           .As<IServiceExistingMethodCapturedArg>();

    // New parameter-supporting method, but with capturing the argument.
    // Name is different just so I can run both side-by-side for now.
    builder.RegisterWithParamSupport((MyDependency1 dep1) => new MyComponent(dep1, "arg"))
           .As<IServiceNewMethodCapturedArg>();

    // New parameter-supporting method, but the argument comes from a resolve parameter.
    // Name is different just so I can run both side-by-side for now.
    builder.RegisterWithParamSupport((MyDependency1 dep1, string arg1) => new MyComponent(dep1, arg1))
           .As<IServiceNewMethodOpenArg>();

    // Check against the reflection way as well.
    builder
       .RegisterType<MyComponent>()
       .UsingConstructor(typeof(MyDependency1), typeof(string))
       .WithParameter(TypedParameter.From("arg"))
       .As<IServiceFromReflection>();

    _container = builder.Build();
}

[Benchmark]
public object ResolveExistingWithCapturedParameter()
{
    return _container.Resolve<IServiceExistingMethodCapturedArg>();
}

[Benchmark]
public object ResolveParamSupportWithCapturedParameter()
{
    return _container.Resolve<IServiceNewMethodCapturedArg>();
}

[Benchmark]
public object ResolveParamSupportWithProvidedTypedParameter()
{
    return _container.Resolve<IServiceNewMethodOpenArg>(TypedParameter.From("arg"));
}

[Benchmark]
public object ResolveParamSupportWithProvidedNamedParameter()
{
    return _container.Resolve<IServiceNewMethodOpenArg>(new NamedParameter("arg1", "value"));
}

[Benchmark]
public object ResolveReflectionWithRegisteredParameter()
{
    return _container.Resolve<IServiceFromReflection>();
}
```

And the results:

```markdown
|                                        Method |       Mean |   Error |  StdDev |  Gen 0 | Allocated |
|---------------------------------------------- |-----------:|--------:|--------:|-------:|----------:|
|          ResolveExistingWithCapturedParameter |   795.0 ns | 2.42 ns | 2.02 ns | 0.2861 |      1 KB |
|      ResolveParamSupportWithCapturedParameter |   795.5 ns | 5.75 ns | 4.49 ns | 0.2861 |      1 KB |
| ResolveParamSupportWithProvidedTypedParameter |   960.8 ns | 3.97 ns | 3.10 ns | 0.3548 |      1 KB |
| ResolveParamSupportWithProvidedNamedParameter |   958.9 ns | 3.21 ns | 3.00 ns | 0.3548 |      1 KB |
|      ResolveReflectionWithRegisteredParameter | 1,009.7 ns | 2.19 ns | 1.83 ns | 0.4025 |      2 KB |
```

As you can see, the new method that supports parameters is no slower than the existing method when the set of parameters is empty.

The downside? It's a good bit more code per-method (holding types, conditional checks, etc).  I'm tempted to add a little private Source Generator to auto-generate the methods for us...I've done a few source generators now, so it wouldn't be particularly tricky.

Reply:
Cool; I'll put a generator together over the weekend.